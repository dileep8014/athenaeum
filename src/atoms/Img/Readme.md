### Img Example:

The `<Img />` component can take also any prop that an HTML5 img can take.

```jsx
    <Img
      src='https://policygenius-blog.imgix.net/2017/09/promo-04.jpg'
    />
```
