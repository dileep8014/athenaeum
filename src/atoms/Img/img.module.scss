.img {
  max-width: 100%;
  width: 100%;
  display: block;
}

.picture {
  width: 100%;
}

@media #{$mobile-only} {
  .hide-mobile-image {
    display: none;
  }
}

@media #{$tablet-only} {
  .hide-tablet-image {
    display: none;
  }
}


// For lazyload
// https://github.com/aFarkas/lazysizes
/* fade image in after load */
/* stylelint-disable selector-pseudo-class-no-unknown */
:global {
  .lazyload,
  .lazyloading {
    opacity: 0;
    width: 100%;
    display: block;
  }

  .unlazy {
    opacity: 1;
  }

  .lazyloaded {
    opacity: 1;
    width: 100%;
    display: block;
    transition: opacity 100ms;
  }
}
