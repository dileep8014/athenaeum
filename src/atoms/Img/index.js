import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './img.module.scss';

if (typeof window !== 'undefined') {
  require('lazysizes/lazysizes.min');
}

const strParams = '&fit=clip&auto=compress&auto=format&ch=Width,DPR,Save-Data';

const imgixSrcStr = (src, width = 768, height = 768) => `https://policygenius-images.imgix.net/${src}` +
  `?w=${Math.min(width, width)}` +
  `&h=${Math.min(height, height)}` +
  '&fit=clip&auto=format&auto=compress';

const imgixSrcset = ( src, maxWidth = Infinity, maxHeight = Infinity ) => `
  https://policygenius-images.imgix.net/${src}?w=${Math.min(maxWidth, 1025)}&h=${Math.min(maxHeight, 1025)}${strParams} 1026w,
  https://policygenius-images.imgix.net/${src}?w=${Math.min(maxWidth, 768)}&h=${Math.min(maxHeight, 768)}${strParams} 769w,
  https://policygenius-images.imgix.net/${src}?w=${Math.min(maxWidth, 320)}&h=${Math.min(maxHeight, 320)}${strParams} 321w
`;

const createName = (str) => {
  let filename = '';

  if (typeof str === 'string') {
    filename = str.match(/([^/]*)(?=\.)/);
  }

  return filename ? filename[0].replace(/(\W+|_)/g, ' ').trim() : 'PolicyGenius';
};

function Img(props) {
  const {
    className,
    src,
    srcSet,
    imgixSrc,
    alt,
    title,
    maxWidth = Infinity,
    maxHeight = Infinity,
    mobileSrc,
    tabletSrc,
    mobileImgixSrc,
    tabletImgixSrc,
    lazy,
    ...rest
  } = props;

  const isPicture = mobileSrc !== undefined ||
    tabletSrc !== undefined ||
    mobileImgixSrc !== undefined ||
    tabletImgixSrc !== undefined;

  const classes = cx(
    styles['img'],
    'lazyload',
    !lazy && 'unlazy',
    className,
  );

  const pictureClasses = cx(
    (mobileImgixSrc === null || mobileSrc === null) && styles['hide-mobile-image'],
    (tabletImgixSrc === null || tabletSrc === null) && styles['hide-tablet-image'],
    className,
  );

  const srcset = (() => {
    if (srcSet) { return srcSet; }
    if (src) { return undefined; }

    return imgixSrcset(`${imgixSrc}`, maxWidth, maxHeight);
  })();

  // TODO: Consider adding a noscript tag to handle a situation where lazysizes doesn't work.
  if (isPicture) {
    const defaultSrc = src || imgixSrcStr(imgixSrc);

    return (
      <picture className={pictureClasses}>
        { ( mobileImgixSrc || mobileSrc ) && <source srcSet={mobileSrc || imgixSrcStr(mobileImgixSrc, 767)} media='(max-width: 767px)' />}
        { ( tabletImgixSrc || tabletSrc ) && <source srcSet={tabletSrc || imgixSrcStr(tabletImgixSrc, 1024)} media='(max-width: 1024px)' />}
        <img
          className={cx(styles.img, 'lazyload')}
          alt={alt || createName(src || imgixSrc)}
          title={title || createName(src || imgixSrc)}
          data-src={defaultSrc}
          data-sizes='auto'
          src={lazy ? undefined : defaultSrc}
          {...rest}
        />
      </picture>
    );
  }

  return (
    <img
      className={classes}
      alt={alt || createName(src || imgixSrc)}
      title={title || createName(src || imgixSrc)}
      data-src={src || imgixSrcStr(imgixSrc)}
      data-srcset={srcset}
      data-sizes={lazy ? 'auto' : '100vw'}
      src={lazy ? undefined : src || imgixSrcStr(imgixSrc)}
      {...rest}
    />
  );
}


Img.propTypes = {
  /**
   * This prop will add a new className to any inherent classNames
   * provided in the component's index.js file.
   */
  className: PropTypes.string,

  /**
   * Recommended: The Imgix base url for the image. Will handle all optomizations automagically.
   * https://docs.imgix.com/apis/url/adjustment/usmrad
   */
  imgixSrc: PropTypes.string,

  /**
   * The src url for the image. If you use this, it will override the Imgix src.
   */
  src: PropTypes.string,

  /**
   * The srcset for the image. If you use this, it will override the Imgix src.
   */
  srcSet: PropTypes.string,

  /**
   * The img alt for the image. Defaults to an improved version of the filename.
   */
  alt: PropTypes.string,

  /**
   * The img title for the image. Defaults to an improved version of the filename.
   */
  title: PropTypes.string,

  /**
   * Used with imgixSrc to prevent larger than necessary images from loading
   */
  maxWidth: PropTypes.string,

  /**
   * Used with imgixSrc to prevent larger than necessary images from loading
   */
  maxHeight: PropTypes.string,

  /**
   * When used, will create a <source> for a <picture> element for mobile.
   * If null, the image will not display on mobile
   */
  mobileSrc: PropTypes.string,
  /**
   * When used, will create a <source> for a <picture> element for tablet
   * If null, the image will not display on tablet
   */
  tabletSrc: PropTypes.string,
  /**
   * Alternate image for mobile. When used, will create a <source> for a <picture> element for mobile. Will use imgix custom string.
   * If null, the image will not display on mobile
   */
  mobileImgixSrc: PropTypes.string,
  /**
   * Alternate image for tablet. When used, will create a <source> for a <picture> element for tablet. Will use imgix. Will use imgix custom string.
   * If null, the image will not display on tablet
   */
  tabletImgixSrc: PropTypes.string,
  /**
   * Set to false to turn off lazyloading
   */
  lazy: PropTypes.bool,
};

Img.defaultProps = {
  alt: '',
  imgixSrc: 'photos/alexander-dummer-150646.jpg',
  lazy: true
};

export default Img;
