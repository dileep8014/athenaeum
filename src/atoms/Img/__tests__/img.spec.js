import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Img from 'atoms/Img';

describe('<Img />', () => {
  const { warn } = console;

  beforeAll(() => {
    console.warn = jest.fn();
  });

  afterAll(() => {
    console.warn = warn;
  });

  it('renders correctly', () => {
    const wrapper = mount(<Img />);

    expect(wrapper.type()).toEqual(Img);
  });

  it('renders correctly with default props', () => {
    const actual = renderer.create(<Img />).toJSON();

    expect(actual).toMatchSnapshot();
  });

  it('renders correctly with given imgix props', () => {
    const imgixProps = {
      className: 'cool-class',
      imgixSrc: 'photos/are-cool.jpg',
      maxWidth: '10',
      maxHeight: '20',
    };

    const actual = renderer.create(<Img {...imgixProps} />).toJSON();

    expect(actual).toMatchSnapshot();
  });

  it('renders correctly with given regular props', () => {
    const regProps = {
      className: 'cool-class',
      src: 'some-url',
      srcSet: 'some-url',
      imgixSrc: 'photos/are-cool.jpg',
      alt: 'alt',
      title: 'title',
      maxWidth: '10',
      maxHeight: '20',
    };

    const actual = renderer.create(<Img {...regProps} />).toJSON();

    expect(actual).toMatchSnapshot();
  });

  it('renders correctly with given regular picture source props', () => {
    const props = {
      alt: 'Alternative',
      mobileSrc: 'small',
      tabletSrc: 'medium',
      desktopSrc: 'large',
      src: 'default.jpg'
    };

    const actual = renderer.create(<Img {...props} />).toJSON();

    expect(actual).toMatchSnapshot();
  });

  it('renders correctly with given imgix picture source props', () => {
    const props = {
      mobileImgixSrc: 'small',
      tabletImgixSrc: 'medium',
      desktopImgixSrc: 'large',
      imgixSrc: 'default.jpg'
    };

    const actual = renderer.create(<Img {...props} />).toJSON();

    expect(actual).toMatchSnapshot();
  });

  it('renders with hide classes when null src props are provided', () => {
    const props = {
      mobileImgixSrc: null,
      tabletImgixSrc: null,
      imgixSrc: 'default.jpg'
    };

    const actual = renderer.create(<Img {...props} />).toJSON();

    expect(actual).toMatchSnapshot();
  });
});
