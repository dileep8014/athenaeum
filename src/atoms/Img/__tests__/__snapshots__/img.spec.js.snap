// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Img /> renders correctly with default props 1`] = `
<img
  alt="alexander dummer 150646"
  className="img lazyload"
  data-sizes="auto"
  data-src="https://policygenius-images.imgix.net/photos/alexander-dummer-150646.jpg?fit=max&auto=format&ch=Width,DPR&w={width}"
  data-srcset="
  https://policygenius-images.imgix.net/photos/alexander-dummer-150646.jpg?fit=max&auto=format&ch=Width,DPR&w={width}&dpr=1 1x,
  https://policygenius-images.imgix.net/photos/alexander-dummer-150646.jpg?fit=max&auto=format&ch=Width,DPR&w={width}&dpr=2 2x,
  https://policygenius-images.imgix.net/photos/alexander-dummer-150646.jpg?fit=max&auto=format&ch=Width,DPR&w={width}&dpr=3 3x,
"
  src={undefined}
  title="alexander dummer 150646"
/>
`;

exports[`<Img /> renders correctly with given imgix picture source props 1`] = `
<picture
  className="picture"
>
  <source
    media="(max-width: 767px)"
    srcSet="https://policygenius-images.imgix.net/small?fit=max&auto=format&ch=Width,DPR&w={width}"
  />
  <source
    media="(max-width: 1024px)"
    srcSet="https://policygenius-images.imgix.net/medium?fit=max&auto=format&ch=Width,DPR&w={width}"
  />
  <img
    alt="default"
    className="img lazyload"
    data-sizes="auto"
    data-src="https://policygenius-images.imgix.net/default.jpg?fit=max&auto=format&ch=Width,DPR&w={width}"
    desktopImgixSrc="large"
    src={undefined}
    title="default"
  />
</picture>
`;

exports[`<Img /> renders correctly with given imgix props 1`] = `
<img
  alt="are cool"
  className="img lazyload cool-class"
  data-sizes="auto"
  data-src="https://policygenius-images.imgix.net/photos/are-cool.jpg?fit=max&auto=format&ch=Width,DPR&w={width}"
  data-srcset="
  https://policygenius-images.imgix.net/photos/are-cool.jpg?fit=max&auto=format&ch=Width,DPR&w={width}&dpr=1 1x,
  https://policygenius-images.imgix.net/photos/are-cool.jpg?fit=max&auto=format&ch=Width,DPR&w={width}&dpr=2 2x,
  https://policygenius-images.imgix.net/photos/are-cool.jpg?fit=max&auto=format&ch=Width,DPR&w={width}&dpr=3 3x,
"
  maxHeight="20"
  maxWidth="10"
  src={undefined}
  title="are cool"
/>
`;

exports[`<Img /> renders correctly with given regular picture source props 1`] = `
<picture
  className="picture"
>
  <source
    media="(max-width: 767px)"
    srcSet="small"
  />
  <source
    media="(max-width: 1024px)"
    srcSet="medium"
  />
  <img
    alt="Alternative"
    className="img lazyload"
    data-sizes="auto"
    data-src="default.jpg"
    desktopSrc="large"
    src={undefined}
    title="default"
  />
</picture>
`;

exports[`<Img /> renders correctly with given regular props 1`] = `
<img
  alt="alt"
  className="img lazyload cool-class"
  data-sizes="auto"
  data-src="some-url"
  data-srcset="some-url"
  maxHeight="20"
  maxWidth="10"
  src={undefined}
  title="title"
/>
`;

exports[`<Img /> renders with hide classes when null src props are provided 1`] = `
<picture
  className="picture hide-mobile-image hide-tablet-image"
>
  <img
    alt="default"
    className="img lazyload"
    data-sizes="auto"
    data-src="https://policygenius-images.imgix.net/default.jpg?fit=max&auto=format&ch=Width,DPR&w={width}"
    src={undefined}
    title="default"
  />
</picture>
`;
